controllers/gatekeeper_controller.go

```diff
@@ -328,6 +330,28 @@ func (r *GatekeeperReconciler) initConfig(ctx context.Context, gatekeeper *opera
 	return nil
 }
 
+func (r *GatekeeperReconciler) StopSubControllers() {
+	log := r.Log.WithName("shutdown")
+	r.StopSubController(log.WithName("Config"), r.configCtrlCtxCancel, &r.isConfigCtrlRunning)
+	r.StopSubController(log.WithName("ConstraintPodStatus"), r.cpsCtrlCtxCancel, &r.isCPSCtrlRunning)
+}
+
+func (r *GatekeeperReconciler) StopSubController(log logr.Logger, cancel context.CancelFunc, running *bool) {
+	if cancel == nil {
+		return
+	}
+
+	log.Info("Stopping the manager.")
+
+	cancel()
+
+	for *running {
```

Comment:
Could we use a sync.WaitGroup instead that is a struct field of `GatekeeperReconciler`? Then every time it starts a "sub" controller, it can add 1 to the wait group. When it stops a controller, it can subtract one from the wait group. Then `StopSubControllers` can call `Wait` on wait group.

Reply:
Ah, I didn't know about that pattern, yes that sounds better--the indefinite loop certainly didn't quite feel right.